src/scripts/eastern_kingdoms/burning_steppes/molten_core/boss_garr.cpp

```diff
@@ -1,200 +1,268 @@
     {
         m_pInstance = static_cast<ScriptedInstance*>(pCreature->GetInstanceData());
-
-        boss_garrAI::Reset();
+        Reset();
     }
 
-    uint32 m_uiAntiMagicPulseTimer;
-    uint32 m_uiMagmaShacklesTimer;
-    uint32 m_uiExplodeTimer;
-    std::list<ObjectGuid> m_lFiresworn;
-
+    EventMap m_CombatEvents;
     ScriptedInstance* m_pInstance;
+    std::vector<ObjectGuid> m_lFiresworn;
 
     void Reset() override
     {
-        m_uiAntiMagicPulseTimer = 15000;
-        m_uiMagmaShacklesTimer  = 10000;
-
-        m_uiExplodeTimer        = 360000; // 6 Minutes
+        m_CombatEvents.Reset();
+        m_lFiresworn.clear();
 
         if (m_creature->IsAlive() && m_pInstance && m_pInstance->GetData(TYPE_GARR) != DONE)
+        {
             m_pInstance->SetData(TYPE_GARR, NOT_STARTED);
+        }
     }
 
     void Aggro(Unit* /*pWho*/) override
     {
-        if (m_pInstance)
+        if (!m_pInstance)
         {
-            if (m_pInstance->GetData(TYPE_GARR) != DONE)
-                m_pInstance->SetData(TYPE_GARR, IN_PROGRESS);
-            else
-            {
-                m_creature->DeleteLater();
-                return;
-            }
+            return;
+        }
+
+        if (m_pInstance->GetData(TYPE_GARR) == DONE)
+        {
+            m_creature->DeleteLater();
+            return;
+        }
 
-            m_creature->SetInCombatWithZone();
+        m_pInstance->SetData(TYPE_GARR, IN_PROGRESS);
+        m_creature->SetInCombatWithZone();
 
-            // count chickens
-            std::list<Creature*> firesworn;
-            GetCreatureListWithEntryInGrid(firesworn, m_creature, NPC_FIRESWORN, 150.0f);
-            m_lFiresworn.clear();
+        // Store add guids
+        std::list<Creature*> adds;
+        GetCreatureListWithEntryInGrid(adds, m_creature, NPC_FIRESWORN, 150.0f);
+        m_lFiresworn.clear();
+        for (const auto& itr : adds)
+        {
+            m_lFiresworn.push_back(itr->GetObjectGuid());
 
-            for (const auto& itr : firesworn)
+            if (Creature* pAdd = m_creature->GetMap()->GetCreature(itr->GetObjectGuid()))
             {
-                m_lFiresworn.push_back(itr->GetObjectGuid());
-
-                if (Creature* pAdd = m_creature->GetMap()->GetCreature(itr->GetObjectGuid()))
-                    DoCastSpellIfCan(pAdd, SPELL_SEPARATION_ANXIETY, CF_TRIGGERED | CF_AURA_NOT_PRESENT);
+                DoCastSpellIfCan(pAdd, SPELL_SEPARATION_ANXIETY, CF_TRIGGERED | CF_AURA_NOT_PRESENT);
             }
         }
+
+        ScheduleCombatEvents();
     }
 
     void JustDied(Unit* /*pKiller*/) override
     {
         if (m_pInstance)
+        {
             m_pInstance->SetData(TYPE_GARR, DONE);
-    }
+        }
 
-    void FireswornJustDied(ObjectGuid fireswornGuid)
-    {
-        m_lFiresworn.remove(fireswornGuid);
-        DoCastSpellIfCan(m_creature, SPELL_ENRAGE, CF_TRIGGERED);
+        // TODO:
+        // When Garr dies, will his adds also die?
```

Comment:
His adds will remain alive when Garr dies, the usual tactic is to kill Garr first and then adds.